modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/disk/ExternalResultHashIndex.java

```diff
@@ -0,0 +1,512 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.query.h2.disk;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Map;
+import org.apache.ignite.IgniteException;
+import org.apache.ignite.internal.processors.cache.persistence.file.FileIO;
+import org.apache.ignite.internal.processors.cache.persistence.file.FileIOFactory;
+import org.apache.ignite.internal.util.typedef.internal.S;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.h2.value.Value;
+import org.h2.value.ValueRow;
+
+import static java.nio.file.StandardOpenOption.CREATE_NEW;
+import static java.nio.file.StandardOpenOption.READ;
+import static java.nio.file.StandardOpenOption.WRITE;
+
+/**
+ * Disk-based hash index for the sorted external result.
+ * Open addressing hashing scheme is used.
+ */
+@SuppressWarnings("TypeMayBeWeakened")
+public class ExternalResultHashIndex implements AutoCloseable {
+    /** Hash map load factor. */
+    private static final double LOAD_FACTOR = 0.5;
+
+    /** Minimum index capacity. */
+    private static final long MIN_CAPACITY = 1L << 8; // 256 entries.
+
+    /** */
+    private static final long MAX_CAPACITY = 1L << 59 - 1;
+
+    /** */
+    private static final long REMOVED_FLAG = 1L << 63;
+
+    /** */
+    private final FileIOFactory fileIOFactory;
+
+    /** Index file directory. */
+    private final String dir;
+
+    /** Row store file directory. */
+    private final String spillFileName;
+
+    /** Current file id. It is incremented on each hashtable extension. */
+    private int id;
+
+    /** Index file. */
+    private File idxFile;
+
+    /** Index file channel. */
+    private FileIO fileIo;
+
+    /** Row store. */
+    private final ExternalResultData rowStore;
+
+    /** Reusable byte buffer for reading and writing. We use this only instance just for prevention GC pressure. */
+    private final ByteBuffer reusableBuff = ByteBuffer.allocate(Entry.ENTRY_BYTES);
+
+    /** Max HashMap size in entries - capacity. */
+    private long cap;
+
+    /** Total count of the stored entries. */
+    private long entriesCnt;
+
+    /**
+     * @param spillFile File with rows.
+     * @param rowStore External result being indexed by this hash index.
+     * @param initSize Init hash map size.
+     */
+    ExternalResultHashIndex(FileIOFactory fileIOFactory, File spillFile, ExternalResultData rowStore, long initSize) {
+        this.fileIOFactory = fileIOFactory;
+        dir = spillFile.getParent();
+        spillFileName = spillFile.getName();
+        this.rowStore = rowStore;
+
+        if (initSize <= MIN_CAPACITY)
+            initSize = MIN_CAPACITY;
+
+        // We need at least the half of hash map be empty to minimize collisions number.
+        long initCap = Long.highestOneBit(initSize) * 2;
```

Comment:
With this code you'll have at least half space being used (not at most). I think you should multiply it by 4 to get what you want.

Reply:
Yes. Good point, thank you. Fixed.